Code/Conviva/Source/Events/ConvivaReporters/AppEventConvivaReporter.swift

```diff
@@ -4,13 +4,40 @@
 
 import ConvivaSDK
 import AVFoundation
+import THEOplayerSDK
 
-struct AppEventConvivaReporter: AppEventProcessor {
+class AppEventConvivaReporter: AppEventProcessor {
+
+    
     let analytics: CISAnalytics
     let videoAnalytics: CISVideoAnalytics
     let adAnalytics: CISAdAnalytics
     let storage: ConvivaConnectorStorage
-
+    
+    init(analytics: CISAnalytics, videoAnalytics: CISVideoAnalytics, adAnalytics: CISAdAnalytics, storage: ConvivaConnectorStorage) {
+        self.analytics = analytics
+        self.videoAnalytics = videoAnalytics
+        self.adAnalytics = adAnalytics
+        self.storage = storage
+    }
+    
+    // This is a workaround for THEOSD-14780
+    // Access log entries for preroll ads come in before the player.ads.playing is updated, which means
+    // bitrate information gets reported to the main content instead, which is incorrect.
+    // Ads get loaded always before access log entries come, which we can use to report
+    // the ad bitrate to the correct endpoint.
+    private var adLoaded: Bool = false
+    private var lastPlayerItem: AVPlayerItem?
+    private var lastAccessLogEvent: AVPlayerItemAccessLogEvent?
+    
+    func adDidLoad(event: THEOplayerSDK.AdLoadedEvent)  {
+        adLoaded = true
+    }
+    func adDidEnd(event: THEOplayerSDK.AdEndEvent)  {
+        adLoaded = false
```

Comment:
maybe it makes sense to reset this also on sourceChange, because it is possible that in the middle of an ad playback a new source is set.

Reply:
Done